### PR TITLE
feat(jsonrpc): add jwt auth middleware

### PR DESCRIPTION
**Pull request by frdomovic**
# feat(jsonrpc): add jwt auth middleware

## Summary
Removed previous auth header for jsonrpc and added new one which takes jwt token from headers, decoded it using jwt secret and passed or denies the request.

Fixes #

## Test plan
Screenshot 2024-08-30 at 13 02 25
Screenshot 2024-08-30 at 13 04 43

### CODE REVIEW

File: crates/server/src/middleware/jwt.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    // `BoxFuture` is a type alias for `Pin<Box<dyn Future + Send + 'a>>`
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request<Body>) -> Self::Future {
+        // todo! experiment with Interior<Store>: WriteLayer<Interior>
+        let result = auth(request.headers(), &self.store.clone());
```
Comment (miraclx) [inline crates/server/src/middleware/jwt.rs]:
```suggestion
        let result = auth(request.headers(), &self.store);
```

File: crates/server/src/middleware/jwt.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            let error_response = err.into_response();
+            return Box::pin(async move { Ok(error_response) });
+        }
+
+        let future = self.inner.call(request);
+
+        Box::pin(async move {
+            let response: Response = future.await?;
+            Ok(response)
+        })
```
Comment (miraclx) [inline crates/server/src/middleware/jwt.rs]:
```suggestion
        Box::pin(self.inner.call(request))
```

File: crates/server/src/middleware/jwt.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    token: String,
+}
+
+pub fn auth(headers: &HeaderMap, store: &Store) -> Result<(), UnauthorizedError<'static>> {
+    let jwt_header = get_jwt_token_from_headers(headers).map_err(|e| {
+        debug!("Failed to extract authentication headers {}", e);
+        UnauthorizedError::new("Failed to extract authentication headers.")
+    })?;
+
+    let jwt_secret = match get_jwt_secret(&store.clone()) {
```
Comment (miraclx) [inline crates/server/src/middleware/jwt.rs]:
```suggestion
    let jwt_secret = match get_jwt_secret(&store) {
```

File: crates/server/src/middleware/jwt.rs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        Ok(Some(secret)) => secret.jwt_secret().to_vec(),
```
Comment (miraclx) [inline crates/server/src/middleware/jwt.rs]:
```suggestion
        Ok(Some(secret)) => secret,
```

or

```suggestion
        Ok(Some(secret)) => *secret.jwt_secret(),
```

File: crates/server/src/middleware/jwt.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        Ok(None) => {
+            return Err(UnauthorizedError::new("JWT secret not found."));
+        }
+        Err(_) => {
+            return Err(UnauthorizedError::new("Failed to fetch JWT secret."));
+        }
+    };
+
+    let _token_data = decode::<Claims>(
```
Comment (miraclx) [inline crates/server/src/middleware/jwt.rs]: we still need to ensure it hasn't expired, yeah?

File: crates/server/src/middleware/jwt.rs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    let auth = JwtHeader {
+        token: token.to_string(),
+    };
+    Ok(auth)
+}
+
+fn extract_token_from_header(authorization_header: &str) -> Result<&str, UnauthorizedError<'_>> {
+    let bearer_prefix = "Bearer ";
+    if authorization_header.starts_with(bearer_prefix) {
+        Ok(&authorization_header[bearer_prefix.len()..])
```
Comment (miraclx) [inline crates/server/src/middleware/jwt.rs]:
```suggestion
    if let Some(token) = authorization_header.strip_prefix("Bearer ") {
        Ok(token)
```

File: crates/server/src/middleware/jwt.rs (hunk 1 above)
Comment (frdomovic) [reply crates/server/src/middleware/jwt.rs]: Updated

File: crates/server/src/middleware/jwt.rs (hunk 2 above)
Comment (frdomovic) [reply crates/server/src/middleware/jwt.rs]: Updated

File: crates/server/src/middleware/jwt.rs (hunk 3 above)
Comment (frdomovic) [reply crates/server/src/middleware/jwt.rs]: Updated

File: crates/server/src/middleware/jwt.rs (hunk 4 above)
Comment (frdomovic) [reply crates/server/src/middleware/jwt.rs]: Updated

File: crates/server/src/middleware/jwt.rs (hunk 5 above)
Comment (frdomovic) [reply crates/server/src/middleware/jwt.rs]: Added check if expired or not

File: crates/server/src/middleware/jwt.rs (hunk 6 above)
Comment (frdomovic) [reply crates/server/src/middleware/jwt.rs]: updated

**Review by miraclx — APPROVED** [review]

**Review by alenmestrov — APPROVED** [review]

**Pull request merged**